### PR TITLE
Log reading capabilities to Ryuko bot

### PR DESCRIPTION
**Pull request by Pixxel123**
This PR adds some log reading functions to the Ryuko bot. Upon uploading of a Ryujinx log file with the format `^Ryujinx_.*\.log$` the bot should be able to read the user's hardware info, operating system and the game name.

In addition, the bot will detect and show which mods are installed, as well as an errors present in the log. The errors shown will be a snippet only, starting from the last error in the log since those can indicate a crash.

If the user meets some criteria a warning will be shown. Currently, these are:

- Having any default logs turned off (Info, Warning, Error, Guest, Stub)
- Having less than 8GB RAM available
- Using macOS
- Using an Intel iGPU on (macOS or Windows)

Here are some previews:

Screenshot 2021-03-05 162143
Screenshot 2021-03-05 162251
Screenshot 2021-03-05 162402
Screenshot 2021-03-05 162445
Screenshot 2021-03-05 162509

The bot now shows the user's controller configuration as well.

I have also implemented restrictions for log analysis to only happen in the #support and #patreon-support channels. Posting a log outside of these channels will show a message warning the user to do so only in those channels.

### CODE REVIEW

**Comment by jms-c** [comment]
Looks good in my opinion.

**Comment by choppymaster** [comment]
I recommend that this feature is enabled only in support channels. Is this available?

**Comment by jms-c** [comment]
> 
> 
> I recommend that this feature is enabled only in support channels. Is this available?

Yes, Commit b4c4438. https://github.com/Ryujinx/ryuko-ng/pull/3/commits/b4c4438b0f64dd90e732277e0f72b0ddabbb8d37

**Comment by jms-c** [comment]
I think this log reading bot is great!

**Comment by choppymaster** [comment]
theOG if you can, can you add an options reader? Example:
Pptc: On
Resolution: 2x 
Docked mode: on
Ignore missing services: off

Like ^ this?

**Comment by Pixxel123** [comment]
> theOG if you can, can you add an options reader? Example:
> Pptc: On
> Resolution: 2x
> Docked mode: on
> Ignore missing services: off

That is the plan, though not all of these are currently readable from the logs. I know PPTC is and I can add that, but most of these are still waiting on https://github.com/Ryujinx/Ryujinx/pull/1942 so that this information can be used.

**Comment by choppymaster** [comment]
The pr is now merged 👍🏻

**Comment by EmulationFanatic** [comment]
> The pr is now merged 👍🏻

No it isn't

File: cogs/logfilereader.py [hunk 1]
```diff
@@ -117,7 +123,12 @@ def format_log_embed():
             )
 
             user_settings_info = "\n".join(
-                (f"**PPTC:** {self.embed['user_settings']['pptc']}",)
+                (
+                    f"**Audio Backend:** `{self.embed['user_settings']['audio_backend']}`",
+                    f"**Switch Mode:** `{self.embed['user_settings']['docked']}`",
```
Comment (choppymaster) [inline cogs/logfilereader.py:128]: How about changing `switch mode` to `console mode`  for better readability?

**Comment by choppymaster** [comment]
> > The pr is now merged 👍🏻
> 
> No it isn't

It is merged, look at that

**Comment by EmulationFanatic** [comment]
> > > The pr is now merged 👍🏻
> > 
> > 
> > No it isn't
> 
> It is merged, look at that

I can't tell if you are new to Github or just trolling, but this PR's status will change from a green "Open" to a purple "Merged" icon when it is merged.
image

**Comment by Pixxel123** [comment]
> 
> 
> > > The pr is now merged 👍🏻
> > 
> > 
> > No it isn't
> 
> It is merged, look at that

I think you mean that Ryujinx/Ryujinx#1942 is merged. In future, to avoid confusion please reference the PR by it's actual reference.

Anyways, I am aware, and I have updated the bot to show the information that the new logging info PR provides. Here's a peek:

Screenshot 2021-03-20 140433

**Comment by choppymaster** [comment]
Oh sorry, thanks for adding this functions.

**Comment by choppymaster** [comment]
> I can't tell if you are new to Github or just trolling, .....
For an info, My account creation was on November 2020 but I know 99% of its function. Trolling is not in my diary.

**Review by choppymaster — APPROVED** [review]
The log reading feature now looks good in my opinion. Thanks!

File: cogs/logfilereader.py (hunk 1 above)
Comment (Pixxel123) [reply cogs/logfilereader.py]: Thanks! Addressed in commit 85c870bf41e7c77a5ebab76fe535c00df0d87ccd

**Comment by CeruleanSky** [comment]
For due diligence, might want to scan the log for roms with filenames that include the name of a pirate sites such as [nswxx], [switch-xxx]. I've seen a few log files with the websites in them before. I know they change all the time, but I'm sure it wouldn't be hard to come up with a top 5 common ones.  Then aborting and alerting them that piracy is not supported.

I know there are some countries were those sites are legal or gray area, but I'd imagine most ryujinx teammembers are not located in them, and probably don't approve of such things regardless.

**Review by Thog — APPROVED** [review]

**Pull request merged**